types/crossfilter2.d.ts

```diff
@@ -0,0 +1,103 @@
+declare namespace CrossFilter2 {
+
+    export interface Selector<T> {
+        (value: T): any;
+    }
+
+    export interface CrossFilterStatic {
+        <T>(data: T[]): CrossFilter<T>;
+        size(): number;
+        version: string;
+        permute<T>(arry: T[], index: number[], deep: number): T[];
+        heap: {
+            <T>(array: T[], lo: number, hi: number): T[];
+            by<T>(value: Selector<T>): Heap<T>;
+        }
+        heapselect: {
+            <T>(array: T[], lo: number, hi: number, k: number): T[];
+            by<T>(value: Selector<T>): HeapSelect<T>;
+        }
+        bisect: {
+            <T>(array: T[], value: T, lo: number, hi: number): number;
+            by<T>(value: Selector<T>): Bisector<T>;
+        }
+        insertionsort: {
+            <T>(array: T[], lo: number, hi: number): T[];
+            by<T>(vaule: Selector<T>): Sort<T>;
+        }
+        quicksort: {
+            <T>(array: T[], lo: number, hi: number): T[];
+            by<T>(value: Selector<T>): Sort<T>;
+        }
+    }
+
+    export interface Bisection<T> {
+        (array: T[], value: T, lo: number, hi: number): number;
+    }
+
+    export interface Bisector<T> extends Bisection<T> {
+        left: Bisection<T>
+        right: Bisection<T>
+    }
+
+    export interface Heap<T> {
+        (array: T[], lo: number, hi: number): T[];
+        sort(array: T[], lo: number, hi: number): T[];
+    }
+
+    export interface HeapSelect<T> {
+        (array: T[], lo: number, hi: number, k: number): T[];
+    }
+
+    export interface Sort<T> {
+        (array: T[], lo: number, hi: number): T[];
+    }
+
+    export interface CrossFilter<T> {
+        add(records: T[]): CrossFilter<T>;
+        remove(predicate?: Function): void;
+        dimension<T>(value: Function, isArray?: boolean): Dimension<T>;
```

Comment:
`value` should be a function from `T` to `number` or `T` to `string` (not `T` to `number|string`).

Reply:
If the answer above is correct (interface of a function that gets passed like in the `remove()` method), then I *think* this would be the correct way to do this.

I'm assuming that the parameter in the value function is `data`.  I'm also not sure how to return `T` to `number` or `T` to `string` than other doing the below.  Maybe, someone with more TypeScript experience can elaborate?

```
export interface ValueFn<T> {
       (data: T): string | number
}

Reply:
You'll have to define 2 interfaces `ValueFnToString` and `ValueFnToNumber` and then you'll have `dimension<T>(value: ValueFnToString | ValueFnToNumber ...`. I think?

Reply:
Hmmm, of course, we also need to allow output of an array of numbers or strings. Wouldn't want it to be simple...